### ISSUE TITLE
Support for writing / rendering math equations

### ISSUE THREAD

**Issue by avikchaudhuri** [excalidraw/excalidraw#1176]
It would be nice to support math equations embedded in text. Imagine the cool diagrams you can draw with labels using this support to illustrate geometry and physics problems.

One way to implement this would be to integrate [LaTeX math mode](https://artofproblemsolving.com/wiki/index.php/LaTeX:Math#Math_Mode) in text. For example, `$d = \sqrt{x^2 + y^2}$` would show up as <img src="https://latex.codecogs.com/gif.latex?d&space;=&space;\sqrt{x^2&space;&plus;&space;y^2}"/>.

Potential challenges:
1. Can this be done in a lightweight manner?
2. How difficult would it be to support artistic rendering in this mode?

**Comment by vjeux** [comment]
@sophiebits suggested looking into https://katex.org/

**Comment by dwelle** [comment]
The biggest hurdle will be to transfer the styles of each character from the katex render over to the canvas:

![image](https://user-images.githubusercontent.com/5153846/78355946-eda51600-75ae-11ea-88d7-c0521b2def03.png)

That looks like a big undertaking.

> Can this be done in a lightweight manner?

I suggest we render math only when delimited with `$$<math here>$$` (although if someone wanted to write literal `$$`, they'd then need to escape `\$\$`). In this way we can do a regex.

> How difficult would it be to support artistic rendering in this mode?

AFAIK we can supply a custom font, but not sure there's a hand-drawn font that supports math symbols.

**Comment by guerby** [comment]
js package for rendering math:

https://www.mathjax.org/

(not tested)

**Comment by guerby** [comment]
Didn't see that mathjax require a server part.

May be this one is simpler :

https://katex.org/

**Comment by jjdredd** [comment]
> How difficult would it be to support artistic rendering in this mode?

I think it isn't that important for latex, especially since math is usually more dense and hard enough to read already.

**Comment by baptiste** [comment]
I'd be keen to see this reopened – Mathjax is a very powerful and mature tool that can produce standalone SVG equations; it feels like a small step to go from there to something Excalidraw can render. It wouldn't be terrible if it was a one-way process (no possibility to edit the original equation once it's become paths) but ideally double-clicking it would allow editing (like normal text) the original LaTeX string. Happy to follow-up on this (I'm currently drawing individual single-stroke glyphs for a similar purpose, and found Mathjax to be the only suitable tool).

**Comment by dwelle** [comment]
Leaving the technical difficulties aside (though when we support #19 it should be easy), I'd also note that unlike your pretty math formulas above, mathjax font style wouldn't be a good fit for the hand-drawn font.

**Comment by baptiste** [comment]
It'd be preferable to have editable paths rather than an embedded image object, in my opinion. The default font isn't ideal I agree but it would be a good start (for the normal and code fonts). For the handwritten font, it's a longer-term prospect because handwritten math fonts aren't really available, but [the project I'm working on](https://tinyurl.com/ba-handwriting) (single stroke math glyphs) would make it quite easy with a bit of roughjs-like path distortion.

**Comment by dwelle** [comment]
It'd still be SVG paths even if in the editor it would act as an image (for the purpose of manipulation) — as for UX, you'd be able to edit the underlying math by double-clicking the shape (or pressing `enter`), same as we with texts today.

**Comment by dwelle** [comment]
There's an open PR for adding support for this. It's not yet clear whether we'll accept this in that form, or at all, but those who want to see this happen give thoughts on Latex vs AsciiMath?

https://github.com/excalidraw/excalidraw/pull/2993

**Comment by RoyiAvital** [comment]
I will be a great addition for technical meeting. I'd go with LaTeX as it is what used widely.

**Comment by baptiste** [comment]
@DanielJGeiger 's PR works wonderfully; I've just extracted all equations from a chapter of lecture slides and imported them – worked a charm.

<img width="1532" alt="Screenshot 2022-01-30 at 20 27 51" src="https://user-images.githubusercontent.com/18970/151714400-d1cd7fc1-6d96-41d9-b9b9-7b9357fe14a4.png">

Unlike SVG images, these can be edited directly in place. I'm planning to create a "library" of equations for each chapter and give that to the students, with the suggestion that they make their own whiteboard "mindmap" organising the course content.

**Comment by baptiste** [comment]
In case anyone else is interested, here's the workflow I used:

1. run [`de-macro`](https://ctan.org/pkg/de-macro) on the original latex file, to replace all custom commands (`\newcommand` and `\def`) with their definition
2. run `pandoc` with a [custom lua filter](https://groups.google.com/g/pandoc-discuss/c/x0_yBMsiJUo/m/AEnWcukEAQAJ) to extract all equations from the latex file
3. [wrap each equation into a JSON object](https://github.com/baptiste/minixcali/blob/main/R/elements.R#L227) that can be imported in Excalidraw (specifically, the PR supporting Mathjax)

The bounding boxes aren't correct until one double-clicks an equation, but that's a trivial enough step to do manually

**Comment by beedrill** [comment]
I do think this is one of the most desired feature for me as a customer. I use Excalidraw mainly as a white board for remote technical team meetings, and being able to quickly write formula is a very important feature. There might also be some good ways to render formula as if they are handwritten:
https://tex.stackexchange.com/questions/453429/create-a-handwriting-environment-for-both-text-and-math

**Comment by baptiste** [comment]
Conversely, I wonder if one could plug into a [Mathpix](https://mathpix.com/ocr) or [deTexify](https://detexify.kirelabs.org/classify.html) APIs to do character recognition from handwritten maths.

**Comment by DanielJGeiger** [comment]
For what it's worth, I closed almost all my PRs other than #2993 hoping to make it easier for @excalidraw to review my work.

@baptiste @jcezarms @mcaruel

**Comment by DanielJGeiger** [comment]
PR #6037 replaces/continues #2993.  Interested parties can use LaTeX (and AsciiMath) at https://math.preview.excalidraw.com.

@baptiste @jcezarms @mcaruel
